app/services/charges/validators/percentage_service.rb

```diff
@@ -0,0 +1,60 @@
+# frozen_string_literal: true
+
+module Charges
+  module Validators
+    class PercentageService < Charges::Validators::BaseService
+      def validate
+        errors = []
+        errors << :invalid_rate unless valid_rate?
+        errors << :invalid_fixed_amount unless valid_fixed_amount?
+        errors << :invalid_fixed_amount_target unless valid_fixed_amount_target?
+
+        return result.fail!(:invalid_properties, errors) if errors.present?
+
+        result
+      end
+
+      private
+
+      def rate
+        properties['rate']
+      end
+
+      def valid_rate?
+        return false unless rate.is_a?(String)
+
+        decimal_amount = BigDecimal(rate)
+
+        decimal_amount.present? && decimal_amount.finite? && decimal_amount.positive?
+      rescue ArgumentError, TypeError
+        false
+      end
+
+      def fixed_amount
+        properties['fixed_amount']
+      end
+
+      def valid_fixed_amount?
+        return false unless fixed_amount.is_a?(String)
+
+        decimal_amount = BigDecimal(fixed_amount)
+
+        decimal_amount.present? &&
+          decimal_amount.finite? &&
+          (decimal_amount.zero? || decimal_amount.positive?)
+      rescue ArgumentError, TypeError
+        false
+      end
+
+      def fixed_amount_target
+        properties['fixed_amount_target']
+      end
+
+      def valid_fixed_amount_target?
+        fixed_amount_target.present? &&
+          fixed_amount_target.is_a?(String) &&
+          (fixed_amount_target == 'all_units' || fixed_amount_target == 'each_unit')
```

Comment:
Maybe the values `all_units` and `each_unit`  could be stored as constants in this class or in the Charge model

```ruby
FIXED_AMOUNT_TARGETS = %w[all_units each_unit].freeze
```